### PR TITLE
Build system rewrite, cleanup, add `make install` and packaging

### PR DESCRIPTION
**Pull request by dev-zero**
The `nvmmConfig/Target.cmake` may still need some tuning, but that will come as I work up my way to OpenFAM itself.
The contained .spec file is being used as base for the RPM package with a slightly more descriptive name, published here: https://build.opensuse.org/package/show/home:dev-zero/openfam-nvmm (will be moved to an OpenCUBE org at some point).

### CODE REVIEW

**Comment by dev-zero** [comment]
with the latest commit build on Fedora Rawhide are succeeding as well

**Comment by gautham-bhat-k** [comment]
Thanks @dev-zero for your contributions. We are currently testing the changes in our environment. We will get back to you soon with our comments.

**Comment by gautham-bhat-k** [comment]
@dev-zero thanks for your contributions. We have tested the code changes in our platform and all tests have passed. But, we see lot of format changes in most of the files,  we usually use git-clang-format with following command :
**git-clang-format --style=file -v --extensions cc,cpp,h** 
So, would it be possible for you to use the same formatting ?

**Comment by dev-zero** [comment]
@gautham-bhat-k thanks for the review. As written in the respective commit 4dfa35359f8458fdf7f9f78510831c5ae412cb53 this was intentional to make code review easier:

```
The convention to run git-clang-format was not in place from the start, resulting
in different formating across the codebase. Introduce a .clang-format which sets
the default indent width to 4 (what is used mostly), based on the LLVM style (which
has been used by git-clang-format implicitly due to the absence of any config).
Add a pre-commit config to reduce manual labour.
```

When merging, I do suggest to not squash the commits, which makes it possible to ignore reformatting requests in `git blame` with one of the following options: `[--ignore-rev <rev>] [--ignore-revs-file <file>]`

File: contrib/docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM opensuse/leap:15.5
```
Comment (rajakrishi) [inline contrib/docker/Dockerfile:1]: We should add copyright here as well

**Review by gautham-bhat-k — APPROVED** [review]

**Review by rajakrishi — CHANGES_REQUESTED** [review]

**Comment by ssinghal53** [comment]
We can do the merge and update the copyright

**Comment by gautham-bhat-k** [comment]
We have also run tests available in this repo, and we do not see any regression from the changes done as part of this pull request. Following log is from regression test :

(base) basriks@gtlogin2:~/openfam-nvmm-new/build> make test
Running tests...
Test project /home/basriks/openfam-nvmm-new/build
      Start  1: ShelfFile.CreateDestroyTruncateExistSize
1/52 Test  #1: ShelfFile.CreateDestroyTruncateExistSize ..........   Passed    0.24 sec
      Start  2: ShelfFile.OpenCloseTruncateExistSize
2/52 Test  #2: ShelfFile.OpenCloseTruncateExistSize ..............   Passed    0.24 sec
      Start  3: ShelfFile.MapUnmap
3/52 Test  #3: ShelfFile.MapUnmap ................................   Passed    0.23 sec
      Start  4: Membership.Create
4/52 Test  #4: Membership.Create .................................   Passed    0.23 sec
      Start  5: Membership.OpenClose
5/52 Test  #5: Membership.OpenClose ..............................   Passed    0.23 sec
      Start  6: Membership.Destroy
6/52 Test  #6: Membership.Destroy ................................   Passed    0.23 sec
      Start  7: Membership.AddRemoveSlots
7/52 Test  #7: Membership.AddRemoveSlots .........................   Passed    0.23 sec
      Start  8: Membership.FindSlots
8/52 Test  #8: Membership.FindSlots ..............................   Passed    0.23 sec
      Start  9: Pool.CreateDestroyExist
9/52 Test  #9: Pool.CreateDestroyExist ...........................   Passed    0.23 sec
      Start 10: Pool.OpenClose
10/52 Test #10: Pool.OpenClose ....................................   Passed    0.23 sec
      Start 11: Pool.ShelfManagement
11/52 Test #11: Pool.ShelfManagement ..............................   Passed    0.23 sec
      Start 12: Pool.ShelfUsage
12/52 Test #12: Pool.ShelfUsage ...................................   Passed    0.23 sec
      Start 13: Pool.MultiThreadStressTest
13/52 Test #13: Pool.MultiThreadStressTest ........................***Exception: SegFault  0.27 sec
      Start 14: Pool.MultiProcessStressTest
14/52 Test #14: Pool.MultiProcessStressTest .......................   Passed   14.96 sec
      Start 15: ShelfRegion.CreateDestroyVerify
15/52 Test #15: ShelfRegion.CreateDestroyVerify ...................   Passed    0.24 sec
      Start 16: ShelfRegion.OpenCloseSize
16/52 Test #16: ShelfRegion.OpenCloseSize .........................   Passed    0.24 sec
      Start 17: ShelfRegion.MapUnmap
17/52 Test #17: ShelfRegion.MapUnmap ..............................   Passed    0.23 sec
      Start 18: FixedBlockAllocator.Test
18/52 Test #18: FixedBlockAllocator.Test ..........................   Passed    0.74 sec
      Start 19: Ownership.Test
19/52 Test #19: Ownership.Test ....................................   Passed    0.25 sec
      Start 20: FreeLists.Test
20/52 Test #20: FreeLists.Test ....................................   Passed    0.30 sec
      Start 21: PoolRegion.CreateDestroyExist
21/52 Test #21: PoolRegion.CreateDestroyExist .....................   Passed    0.25 sec
      Start 22: PoolRegion.OpenCloseSize
22/52 Test #22: PoolRegion.OpenCloseSize ..........................   Passed    0.24 sec
      Start 23: PoolRegion.MapUnmap
23/52 Test #23: PoolRegion.MapUnmap ...............................   Passed    0.24 sec
      Start 24: EpochZoneHeap.Free
24/52 Test #24: EpochZoneHeap.Free ................................   Passed    0.30 sec
      Start 25: EpochZoneHeap.DelayedFree
25/52 Test #25: EpochZoneHeap.DelayedFree .........................   Passed    1.31 sec
      Start 26: EpochZoneHeap.Resize
26/52 Test #26: EpochZoneHeap.Resize ..............................   Passed    0.48 sec
      Start 27: EpochZoneHeap.MultipleResize
27/52 Test #27: EpochZoneHeap.MultipleResize ......................   Passed    4.34 sec
      Start 28: EpochZoneHeap.MultipleResizeBoundary
28/52 Test #28: EpochZoneHeap.MultipleResizeBoundary ..............   Passed    0.33 sec
      Start 29: EpochZoneHeap.MultipleResizeBoundaryFail
29/52 Test #29: EpochZoneHeap.MultipleResizeBoundaryFail ..........   Passed    0.32 sec
      Start 30: EpochZoneHeap.SmallerResize
30/52 Test #30: EpochZoneHeap.SmallerResize .......................   Passed    0.32 sec
      Start 31: EpochZoneHeap.PowerOfTwoResize
31/52 Test #31: EpochZoneHeap.PowerOfTwoResize ....................   Passed    0.33 sec
      Start 32: EpochZoneHeap.OffsetAllocResize
32/52 Test #32: EpochZoneHeap.OffsetAllocResize ...................   Passed    0.34 sec
      Start 33: EpochZoneHeap.AllocResize
33/52 Test #33: EpochZoneHeap.AllocResize .........................   Passed    0.36 sec
      Start 34: EpochZoneHeap.DelayedFreeResize
34/52 Test #34: EpochZoneHeap.DelayedFreeResize ...................   Passed    2.35 sec
      Start 35: EpochZoneHeap.MultipleDelayedFreeResize
35/52 Test #35: EpochZoneHeap.MultipleDelayedFreeResize ...........   Passed    3.83 sec
      Start 36: EpochZoneHeap.MultipleDelayedFreeResizeClose
36/52 Test #36: EpochZoneHeap.MultipleDelayedFreeResizeClose ......   Passed    4.33 sec
      Start 37: EpochZoneHeap.Permissions
37/52 Test #37: EpochZoneHeap.Permissions .........................   Passed    0.31 sec
      Start 38: EpochZoneHeap.Merge
38/52 Test #38: EpochZoneHeap.Merge ...............................   Passed    0.38 sec
      Start 39: EpochZoneHeap.LargePoolId
39/52 Test #39: EpochZoneHeap.LargePoolId .........................   Passed    0.57 sec
      Start 40: EpochZoneHeap.Largeallocsize
40/52 Test #40: EpochZoneHeap.Largeallocsize ......................   Passed    0.30 sec
      Start 41: EpochZoneHeap.MergeAllocFree
41/52 Test #41: EpochZoneHeap.MergeAllocFree ......................   Passed    1.56 sec
      Start 42: EpochZoneHeap.NoDelayedFree
42/52 Test #42: EpochZoneHeap.NoDelayedFree .......................   Passed    1.28 sec
      Start 43: EpochZoneHeapResize.MultiProcessResize
43/52 Test #43: EpochZoneHeapResize.MultiProcessResize ............   Passed    8.32 sec
      Start 44: MemoryManagerDeathTest.HeapWithMapUnmapReadOnly
44/52 Test #44: MemoryManagerDeathTest.HeapWithMapUnmapReadOnly ...   Passed    0.36 sec
      Start 45: MemoryManager.Region
45/52 Test #45: MemoryManager.Region ..............................   Passed    0.24 sec
      Start 46: MemoryManager.Heap
46/52 Test #46: MemoryManager.Heap ................................   Passed    0.36 sec
      Start 47: MemoryManager.HeapWithMapUnmapPointer
47/52 Test #47: MemoryManager.HeapWithMapUnmapPointer .............   Passed    0.36 sec
      Start 48: MemoryManager.HeapWithMapUnmapReadWrite
48/52 Test #48: MemoryManager.HeapWithMapUnmapReadWrite ...........   Passed    0.36 sec
      Start 49: MemoryManager.HeapWithGlobalLocalPtr
49/52 Test #49: MemoryManager.HeapWithGlobalLocalPtr ..............   Passed    0.36 sec
      Start 50: MemoryManager.HeapHugeObjects
50/52 Test #50: MemoryManager.HeapHugeObjects .....................   Passed    0.36 sec
      Start 51: MemoryManager.MultiThreadStressTest
51/52 Test #51: MemoryManager.MultiThreadStressTest ...............   Passed    0.26 sec
      Start 52: MemoryManager.MultiProcessHeap
52/52 Test #52: MemoryManager.MultiProcessHeap ....................   Passed    0.44 sec
 
98% tests passed, 1 tests failed out of 52
 
Total Test time (real) =  55.64 sec
 
The following tests FAILED:
         13 - Pool.MultiThreadStressTest (SEGFAULT)
Errors while running CTest
Output from these tests are in: /home/basriks/openfam-nvmm-new/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
make: *** [Makefile:105: test] Error 8
(base) basriks@gtlogin2:~/openfam-nvmm-new/build>
The Pool.MultiThreadStressTest (SEGFAULT) is expected failure due to program environment

**Review by rajakrishi — APPROVED** [review]
Approving

**Pull request merged**